examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+             || (std::fabs(p[1] + 1) < 1e-8 && p[0] >= 0.5) // y == -1, x > 0.5
+    )
+      {
+        return 1.0;
+      }
+    else
+      {
+        return 0.0;
+      }
+  }
+
+
+
+  template <int dim>
+  void BoundaryValues<dim>::value_list(const std::vector<Point<dim>> &points,
+                                       std::vector<double> &          values,
+                                       const unsigned int component) const
+  {
+    Assert(values.size() == points.size(),
+           ExcDimensionMismatch(values.size(), points.size()));
+
+    for (unsigned int i = 0; i < points.size(); ++i)
+      values[i] = BoundaryValues<dim>::value(points[i], component);
+  }
+
+  template <int dim>
+  double delta_value(const double         hk,
+                     const double         eps,
+                     const Tensor<1, dim> dir,
+                     const double         pk)
+  {
+    // Value defined in 'On discontinuity–capturing methods for
+    // convection–diffusion equations'
+    double Peclet = dir.norm() * hk / (2.0 * eps * pk);
+    double coth =
+      (1.0 + std::exp(-2.0 * Peclet)) / (1.0 - std::exp(-2.0 * Peclet));
+
+    return hk / (2.0 * dir.norm() * pk) * (coth - 1.0 / Peclet);
+  }
+
+
+  template <int dim>
+  class AdvectionProblem
+  {
+  public:
+    AdvectionProblem(Settings settings);
+    void run();
+
+  private:
+    void setup_system();
+
+    template <class IteratorType>
+    void assemble_cell(const IteratorType &cell,
+                       ScratchData<dim> &  scratch_data,
+                       CopyData &          copy_data);
+    void assemble_system_and_multigrid();
+
+    std::unique_ptr<MGSmoother<Vector<double>>> create_smoother();
+
+    void solve();
+    void refine_grid();
+    void output_results(const unsigned int cycle) const;
+
+    Triangulation<dim> triangulation;
+    DoFHandler<dim>    dof_handler;
+
+    FE_Q<dim>     fe;
+    MappingQ<dim> mapping;
+    unsigned int  quad_degree;
+
+    AffineConstraints<double> constraints;
+
+    SparsityPattern      sparsity_pattern;
+    SparseMatrix<double> system_matrix;
+
+    Vector<double> solution;
+    Vector<double> system_rhs;
+
+    MGLevelObject<SparsityPattern> mg_sparsity_patterns;
+    MGLevelObject<SparsityPattern> mg_interface_sparsity_patterns;
+
+    MGLevelObject<SparseMatrix<double>> mg_matrices;
+    MGLevelObject<SparseMatrix<double>> mg_interface_in;
+    MGLevelObject<SparseMatrix<double>> mg_interface_out;
+
+    MGConstrainedDoFs mg_constrained_dofs;
+
+    Settings       settings;
+    const double   epsilon;
+    Tensor<1, dim> advection_direction;
+  };
+
+
+
+  template <int dim>
+  AdvectionProblem<dim>::AdvectionProblem(Settings settings)
+    : triangulation(Triangulation<dim>::limit_level_difference_at_vertices)
+    , dof_handler(triangulation)
+    , fe(settings.fe_degree)
+    , mapping(settings.fe_degree)
+    , quad_degree(2 * fe.degree + 2)
+    , settings(settings)
+    , epsilon(0.005)
```

Comment:
yes, indeed several of these could be made `const`.

Since the `settings` variable is different from the rest conceptually, I'd make it first or last among the member variables.